lib/models/watch-detector.js

```diff
@@ -0,0 +1,238 @@
+      this.fs.unwatch(tmpDir);
+    } catch (e) {
+      debug('testing if node watcher failed with: %o', e)
+      resolve(false);
+      return;
+    } finally {
+      try {
+        quickTemp.remove(this, 'tmp');
+        // cleanup dir
+      } catch (e) {
+        // not much we can do, lets debug log.
+        debug('cleaning up dir failed with: %o', e)
+      }
+    }
+
+    // it seems we where able to at least watch and unwatch, this should catch
+    // systems that have a very broken NodeWatcher.
+    //
+    // NOTE: we could also add a more advance chance, that triggers a change
+    // and expects to be informed of the change. This can be added in a future
+    // iteration.
+    //
+
+    this._doesNodeWorkCache[key] = true;
+    resolve(true);
+  }.bind(this));
+}
+
+/*
+ * @public
+ *
+ * Selecting the best watcher is complicated, this method (given a preference)
+ * will test and provide the best possible watcher available to the current
+ * system and project.
+ *
+ *
+ *  @method findBestWatcherOption
+ *  @input {Object} options
+ *  @returns {Object} watch preference
+*/
+WatchDetector.prototype.findBestWatcherOption = function(options) {
+  var preference = this.extractPreferenceFromOptions(options);
+  var original = options.watcher;
+  if (preference.watcher === WATCHMAN) {
+    preference = this.checkWatchman();
+  }
+
+  return Promise.resolve(preference).then(function(bestOption) {
+    if (bestOption.watcher === NODE) {
+      // although up to this point, we may believe Node is the best watcher
+      // this may not be true because:
+      // * not all platforms that run node, support node.watch
+      // * not all file systems support node watch
+      //
+      return this.testIfNodeWatcherAppearsToWork().then(function(appearsToWork) {
+        return new WatchPreference(appearsToWork ? NODE : POLLING);
+      });
+    } else {
+      return bestOption;
+    }
+  }.bind(this)).then(function(actual) {
+    debug('foundBestWatcherOption, preference was: %o, actual: %o', options, actual);
+    if (actual /* if no preference was initial set, don't bother informing the user */ &&
+      original /* if no original was set, the fallback is expected */ &&
+      actual.watcher !== original) {
+      // if there was an initial preference, but we had to fall back inform the user.
+      this.ui.writeLine('was unable to use: "' + original + '", fell back to: "' + actual.watcher + '"')
+    }
+    return actual;
+  }.bind(this));
+}
+
+/*
+ * @public
+ *
+ * Although watchman may be selected, it may not work due to:
+ *
+ *  * invalid version
+ *  * it may be broken in detectable ways
+ *  * watchman executable may be something unexpected
+ *  * ???
+ *
+ *  @method checkWatchman
+ *  @returns {Object} watch preference + _watchmanInfo
+*/
+WatchDetector.prototype.checkWatchman = function() {
+  var result = new WatchPreference(null);
+
+  return this.childProcess.exec('watchman version').then(function(output) {
```

Comment:
Might want a `debug` for this check, and its result (including the error case when non-JSON is emitted)